Fix Abs path validation on Windows

mxpv opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind regression
/kind failing-test


#### What this PR does / why we need it:
Fixes broken unit test on windows after https://github.com/kubernetes/kubernetes/pull/112957

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes https://github.com/kubernetes/kubernetes/issues/124076

#### Special notes for your reviewer:
Slack discussion: https://kubernetes.slack.com/archives/C0SJ4AFB7/p1711553979064829

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
none
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
none
```

claudiubelu committed:
Windows: Consider slash-prefixed paths as absolute

filepath.IsAbs does not consider "/" or "\" as absolute paths, even
though files can be addressed as such. [1][2]

Currently, there are some unit tests that are failing on Windows due to
this reason.

[1] https://learn.microsoft.com/en-us/dotnet/standard/io/file-path-formats#traditional-dos-paths
[2] https://learn.microsoft.com/en-us/windows/win32/fileio/naming-a-file#fully-qualified-vs-relative-paths

mxpv commented:
/test pull-ci-kubernetes-unit-windows

mxpv commented:
/test pull-ci-kubernetes-unit-windows

jsturtevant commented:
/assign @claudiubelu 
/sig windows

aravindhp commented:
/priority critical-urgent
/triage accepted

jsturtevant commented on pkg/kubelet/apis/config/mutate_windows.go:
```diff
@@ -0,0 +1,65 @@
+	// in https://github.com/kubernetes/kubernetes/blob/60d73ba75148483d150622a4d8b6e122e189ceab/cmd/kubeadm/app/componentconfigs/kubelet_windows.go#L31
+	klog.V(2).Infoln("Adapting the paths in the KubeletConfiguration for Windows...")
+
+	// Get the drive from where the kubeadm binary was called.
```
```suggestion
	// Get the drive from where the kubelet binary was called.
```

SergeyKanzhelev commented:
/milestone v1.30

SergeyKanzhelev commented on pkg/kubelet/apis/config/mutate_windows.go:
```diff
@@ -0,0 +1,65 @@
+	// Get the drive from where the kubeadm binary was called.
+	exe, err := os.Executable()
+	if err != nil {
+		return errors.Wrap(err, "could not obtain information about the kubeadm executable")
```
```suggestion
		return errors.Wrap(err, "could not obtain information about the kubelet executable")
```

mxpv replied:
Resolved in https://github.com/kubernetes/kubernetes/pull/124084/commits/ab2b3cd8193b4245aa733eaa78624367371e3590

jsturtevant replied:
I think something with the squash went wrong here

mxpv replied:
Hmm. yup, it didn't squash the commit accepted on GitHub, weird. Fixed and resquashed.

SergeyKanzhelev commented on cmd/kubelet/app/options/options.go:
```diff
@@ -212,6 +212,9 @@ func NewKubeletConfiguration() (*kubeletconfig.KubeletConfiguration, error) {
 		return nil, err
 	}
 	applyLegacyDefaults(config)
+	if err := config.Mutate(); err != nil {
```
@haircommander @sohankunkerkar just checking, this will not interfere with the new drop-in configuration feature? 

I don't think there are any reasons to do the mutation AFTER all the merges. If customer applied a bad config, it is their responsibility. Just checking in case you may think of a problem here

SergeyKanzhelev commented on pkg/kubelet/apis/config/mutate_windows.go:
```diff
@@ -0,0 +1,65 @@
+
+func (kc *KubeletConfiguration) Mutate() error {
+	// This replicates configuration fix for Windows originally introduced
+	// in https://github.com/kubernetes/kubernetes/blob/60d73ba75148483d150622a4d8b6e122e189ceab/cmd/kubeadm/app/componentconfigs/kubelet_windows.go#L31
```
I wonder if the fixup of other paths will be needed in future. I think it will be a good practice. Perhaps add a TODO for those so we can do it next relese?

mxpv replied:
Added in https://github.com/kubernetes/kubernetes/pull/124084/commits/1d12452f9209251f29f4b00a383b762cf327336b

aravindhp replied:
It will also help if an issue is created for this. It will be a good first issue.

SergeyKanzhelev commented on pkg/kubelet/apis/config/mutate_windows.go:
```diff
@@ -0,0 +1,65 @@
+	}
+
+	drive := filepath.VolumeName(filepath.Dir(exe))
+	klog.V(2).Infof("Assuming Windows drive %q", drive)
```
so this log message will be written on EVERY Windows start since we mutating the defaul that is guaranteed to have this. Is it desireable?

mxpv replied:
Fixed https://github.com/kubernetes/kubernetes/pull/124084/commits/58c88c8abe90fd3ab9bc245b8df71c8b447cc53c

SergeyKanzhelev commented on pkg/kubelet/apis/config/mutate_windows.go:
```diff
@@ -0,0 +1,65 @@
+
+		// Prepend the drive letter to the path and update the field.
+		*field = filepath.Join(drive, *field)
+		klog.V(2).Infof("kubelet/Windows: adapted path for field %q to %q", name, *field)
```
same comment as above. It feels to me that mutate with logs is logical if we apply it after reading the configuration. In this case we are doing it before reading user-supplied config, so log messages will mostly be noise

mxpv replied:
https://github.com/kubernetes/kubernetes/pull/124084/commits/58c88c8abe90fd3ab9bc245b8df71c8b447cc53c

SergeyKanzhelev commented:
it feels to me that if we are only fixing up defaults, than conditional logic on setting those defaults would be better. But this approach works as well.

Can you please squash commits, I think it is ready to merge

mxpv commented:
> Can you please squash commits, I think it is ready to merge

Done!

SergeyKanzhelev commented on pkg/kubelet/apis/config/mutate_windows.go:
```diff
@@ -0,0 +1,64 @@
+	drive := filepath.VolumeName(filepath.Dir(exe))
+
+	// Mutate the paths in the config.
+	mutatePaths(kc, drive)
```
@jsturtevant, @claudiubelu is it what happens today with the default path? Or we are taking the system drive? Or those must be the same?

jsturtevant replied:
I don't know that we are looking up the drive path in other code paths.  Typically `/path` is written to the current system drive, with out intervention.  It seems that the offending error is `isAbs` isn't recognizing `/path` correctly.

neolit123 reviewed (commented):
@SergeyKanzhelev @mxpv @aravindhp 

this is copied from kubeadm
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/componentconfigs/kubelet_windows.go

in the case of kubeadm there is a single central config uploaded to the cluster then when a windows node joins it must mutate some paths or the kubelet on the windows node will fail. there is a better explanation in the above GO file. it is essentially workarounding a problem in the kubelet / client-go with Windows paths that cannot be easily patched. IIUC and at this point.

i don't think this PR is correct to mutate values and assume the drive letter.
instead it should have separate ways to validate the field on Linux and Windows.
on Linux it can use the stdlib isAbs and on Windows it can use a custom isAbs that passes as "true" without a drive letter in the path string.

using a path such as "/some/dir" on Windows would just write to "CurrentProcessDrive:\\some\\dir" which is fine. the mutation here seems redudant.

haircommander commented on pkg/kubelet/apis/config/mutate_windows.go:
```diff
@@ -0,0 +1,64 @@
+		// path.IsAbs() is not reliable here in the Windows runtime, so check if the
+		// path starts with "/" instead. This means the path originated from a Unix node and
+		// is an absolute path.
+		if !strings.HasPrefix(*field, "/") {
```
I'm a bit worried that this is all of the validation we're doing to find a path, though I concede it seems to be good enough for now. I've manually looked through the kubelet config options and can't find any that should start with `/` but are not a path. That doesn't guarantee there won't be one in the future though.

liggitt replied:
this is really hard to reason about ... is the root cause of the issue on Windows in this function:

```
// resolveRelativePaths makes relative paths absolute by resolving them against `root`
func resolveRelativePaths(paths []*string, root string) {
	for _, path := range paths {
		// leave empty paths alone, "no path" is a valid input
		// do not attempt to resolve paths that are already absolute
		if len(*path) > 0 && !filepath.IsAbs(*path) {
			*path = filepath.Join(root, *path)
		}
	}
}
```

should we fix that there instead of mutating config further?

liggitt replied:
I'm unsure of the workaround in cmd/kubeadm/app/componentconfigs/kubelet_windows.go making absolute unix paths relative to the kubelet binary path instead of the drive of the working dir.

I don't think we should spread that workaround into the kubelet itself, but should fix resolveRelativePaths to do the right thing on Windows.

jsturtevant replied:
the root cause is specifically `ilepath.IsAbs` on windows doesn't respective paths that start with slashes. 
 I think something like https://github.com/kubernetes/kubernetes/pull/112566 should be used.  

The other part of this is that the resolveRelativePaths isn't called when only command line parameters are passed (https://github.com/kubernetes/kubernetes/issues/124076#issuecomment-2032417791)

neolit123 replied:
> I'm unsure of the workaround in cmd/kubeadm/app/componentconfigs/kubelet_windows.go making absolute unix paths relative to the kubelet binary path instead of the drive of the working dir.

+1 i commented the same.

kannon92 commented:
/test

kannon92 commented:
/test pull-ci-kubernetes-unit-windows

aravindhp commented:
@neolit123 
 
> i don't think this PR is correct to mutate values and assume the drive letter. instead it should have separate ways to validate the field on Linux and Windows. on Linux it can use the stdlib isAbs and on Windows it can use a custom isAbs that passes as "true" without a drive letter in the path string.
> 

I am fine with this approach too.

> using a path such as "/some/dir" on Windows would just write to "CurrentProcessDrive:\some\dir" which is fine. the mutation here seems redudant.

Agree.

jsturtevant commented:
> instead it should have separate ways to validate the field on Linux and Windows.
> on Linux it can use the stdlib isAbs and on Windows it can use a custom isAbs that passes as "true" without a drive letter in the path string.

We had a previous PR that went stale that was approved and would have done this https://github.com/kubernetes/kubernetes/pull/112566

aravindhp commented:
> We had a previous PR that went stale that was approved and would have done this #112566

Yup, this is exactly what it would look like

neolit123 commented:
the change in https://github.com/kubernetes/kubernetes/pull/112566 seems OK, but it's missing unit tests.
this PR can be adapted to use that diff + add unit tests.

alternatively as a quick 1.30 fix this PR can branch _windows.go and _unix.go files for this validation:
```
if kc.PodLogsDir == "" {
		allErrors = append(allErrors, fmt.Errorf("invalid configuration: podLogsDir was not specified"))
	}

	if !filepath.IsAbs(kc.PodLogsDir) {
		allErrors = append(allErrors, fmt.Errorf("invalid configuration: pod logs path %q must be absolute path", kc.PodLogsDir))
	}

	if filepath.Clean(kc.PodLogsDir) != kc.PodLogsDir {
		allErrors = append(allErrors, fmt.Errorf("invalid configuration: pod logs path %q must be normalized", kc.PodLogsDir))
```

in a new function `validatePodLogsDir`. without the IsAbs on the Windows variant.
there is already validation_windows.go

https://github.com/mxpv/kubernetes/blob/83751630081e5180b633aea7560d689246780582/pkg/kubelet/apis/config/validation/validation_windows.go

mxpv commented:
I'll take a look.

jsturtevant commented:
> the change in #112566 seems OK, but it's missing unit tests. this PR can be adapted to use that diff + add unit tests.
> 
> alternatively as a quick 1.30 fix this PR can branch _windows.go and _unix.go files for this validation:

I was testing this out and think we need to use something like https://github.com/kubernetes/kubernetes/pull/112566, otherwise the default is moved to the path where the config is loaded (https://github.com/kubernetes/kubernetes/issues/124076#issuecomment-2026149143) becuase of the way it is parsed in 

https://github.com/kubernetes/kubernetes/blob/f4e246bc93ffb68b33ed67c7896c379efa4207e7/pkg/kubelet/kubeletconfig/configfiles/configfiles.go#L98-L107

jsturtevant commented:
before https://github.com/kubernetes/kubernetes/pull/112957

```
PS C:\var\log> ls


    Directory: C:\var\log


Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
d-----         3/28/2024  10:05 PM                containers
d-----         3/28/2024  10:05 PM                pods

```

after 

```
PS C:\var\log> ls


    Directory: C:\var\log


Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
d-----         3/28/2024  10:03 PM                containers
d-----         3/28/2024   9:45 PM                kubelet
```

neolit123 commented:
> resolveRelativePaths

or update this to handle paths properly on windows.
and not treat `/foo/bar` as a relative path.

pacoxu commented:
Any update. 

We will have a release cut today. Is this a release cut blocker? Or just v1.30 final release blocker?

neoaggelos commented:
Hi all, release team here, representing the Release Signal team for 1.30. Can someone reply whether this is considered a release blocker for the v1.30.0-rc.1, or can wait for the v1.30.0 release? The 1.30.0-rc.1 cut is currently blocked because of this.

cc @mxpv @SergeyKanzhelev @marosset

jsturtevant commented:
discussed in [slack](https://kubernetes.slack.com/archives/C0SJ4AFB7/p1712051407401399), this needn't be a blocker for RC1 it definitely should be a blocker for RC2 and the release

mxpv committed:
Add test to verify IsAbs for windows

Signed-off-by: Maksym Pavlenko <pavlenko.maksym@gmail.com>

mxpv committed:
Fix abs path validation on windows

Signed-off-by: Maksym Pavlenko <pavlenko.maksym@gmail.com>

mxpv commented:
> the change in https://github.com/kubernetes/kubernetes/pull/112566 seems OK, but it's missing unit tests.
this PR can be adapted to use that diff + add unit tests.

Ok, so I've reworked this PR:
- Dropped the original approach with mutations
- Cherry-picked https://github.com/kubernetes/kubernetes/pull/112566
- Added a unit test and updated the config validation logic

mxpv commented:
/test pull-ci-kubernetes-unit-windows

liggitt commented on pkg/util/filesystem/util_unix.go:
```diff
@@ -35,3 +36,6 @@ func IsUnixDomainSocket(filePath string) (bool, error) {
 }
+
+// IsAbs is filepath.IsAbs.
+var IsAbs = filepath.IsAbs
```
nit, slight preference for `func IsAbs(path string) bool { return filepath.IsAbs(path) }` to avoid mutable package var

mxpv replied:
Addressed

liggitt commented on pkg/util/filesystem/util_windows.go:
```diff
@@ -85,3 +86,10 @@ func IsUnixDomainSocket(filePath string) (bool, error) {
+// On Windows, filepath.IsAbs will not return True for paths prefixed with a slash, even
+// though they can be used as absolute paths (https://docs.microsoft.com/en-us/dotnet/standard/io/file-path-formats).
+func IsAbs(path string) bool {
+	return filepath.IsAbs(path) || (len(path) > 0 && (path[0] == '\\' || path[0] == '/'))
```
slight preference for this for readability:

```suggestion
	return filepath.IsAbs(path) || strings.HasPrefix(path, `/`) || strings.HasPrefix(path, `\`)
```

liggitt replied:
is a simple check for a `/` or `\` prefix sufficient? does that open us up to tolerating problematic things like https://github.com/golang/go/issues/63713?

liggitt replied:
is `!filepath.IsLocal` closer to what we want for windows?

mxpv replied:
> slight preference for this for readability:

Fixed

> is !filepath.IsLocal closer to what we want for windows?

@jsturtevant do you have any thoughts?

marosset replied:
I believe on Windows checking for `\` or `/` should be sufficient.

Here is a good summary of all the different path formats you can use on Windows
https://learn.microsoft.com/en-us/dotnet/standard/io/file-path-formats

One formatting outlier is something like this example
> `C:Projects\apilibrary\apilibrary.sln`	A relative path from the current directory of the C: drive.

This is acutally using a local path from the current working directory on the `C:` drive (see the note under https://learn.microsoft.com/en-us/dotnet/standard/io/file-path-formats#traditional-dos-paths)

Should we try and handle that?

jsturtevant replied:
> is a simple check for a `/` or `\` prefix sufficient? does that open us up to tolerating problematic things like [golang/go#63713](https://github.com/golang/go/issues/63713)?

reading through the change set for that, I see:

```
IsAbs now correctly reports paths beginning with \??\
as absolute.
```

which would match the check we are performing.  It seems the functions that manipulated these were problematic, not that the user was using them.   Since this isn't manipulating and we still use filepath for the manipulation I think we are ok. 

The other consideration would be that this is not a dynamic value and will be set by the cluster operator and not changeable by users if understand correctly.

jsturtevant replied:
> One formatting outlier is something like this example
> 
> > `C:Projects\apilibrary\apilibrary.sln`	A relative path from the current directory of the C: drive.
> 
> This is acutally using a local path from the current working directory on the `C:` drive (see the note under https://learn.microsoft.com/en-us/dotnet/standard/io/file-path-formats#traditional-dos-paths) (TBH I didn't know about this until reading through this article)
> 
> Should we try and handle that?

I would expect `IsAbs` to handle it?  It doesn't start with `/` or `\` so the added logic wouldn't identify it as absolute

jsturtevant replied:
> The other consideration would be that this is not a dynamic value and will be set by the cluster operator and not changeable by users if understand correctly.

I see we are using this in other places too, that might be dynamic.  Still, from my understanding, I think that the problematic functions where the path manipulations not the fact that it started with `/`.  

This does lead me to question if we should we scope the change for this fix to the bug (in the kubelet configuration) that is blocking 1.30rc2?

liggitt replied:
yes, please scope the change to the kubelet configuration (resolveRelativePaths and ValidateKubeletConfiguration)

liggitt replied:
> is a simple check for a `/` or `\` prefix sufficient? does that open us up to tolerating problematic things...

I have _such_ a hard time understanding windows file path behavior in isolation, and in combination with go's layer on top of it. I don't want to open this up to allowing reserved names and network paths if we can help it. Can we make this fix as narrow as possible to accommodate the unix-style absolute path that is causing the problem, without opening us up to weird things about network paths (`//...`), etc?

Something like `filepath.IsAbs(path) || (strings.HasPrefix(path, `/`) && !strings.HasPrefix(path, `//`))`, perhaps?

jsturtevant commented:
/retest

marosset commented:
/test pull-kubernetes-e2e-capz-windows-master

marosset commented:
The cluster is not coming up in `pull-kubernetes-e2e-capz-windows-master` 
It looks some some of the expected containers on the linux control-plane node aren't starting for some reason.

marosset commented:
We are not seeing the same issue in periodic jobs against master / release-1.30
Let me see if I can get some logs here

marosset commented:
I'm seeing this error in the kubelet log for the Windows machines

> E0403 22:29:44.976846    4488 run.go:74] "command failed" err="failed to validate kubelet configuration, error: invalid configuration: pod logs path \"/var/log/pods\" must be normalized, path: &TypeMeta{Kind:,APIVersion:,}"

Kubelet logs are at https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/124084/pull-kubernetes-e2e-capz-windows-master/1775621190912577536/artifacts/clusters/capz-conf-nrvjly/machines/capz-conf-nrvjly-md-win-nb54n-c5ptz/kubelet.log

mxpv commented:
I guess we should skip that check on windows as well.
Let me fix that.

mxpv committed:
Skipp path clean check for podLogDir on windows

Signed-off-by: Maksym Pavlenko <pavlenko.maksym@gmail.com>

aravindhp commented:
/test pull-kubernetes-e2e-capz-windows-master

marosset commented:
/lgtm

claudiubelu commented on pkg/util/filesystem/util_unix.go:
```diff
@@ -35,3 +36,8 @@ func IsUnixDomainSocket(filePath string) (bool, error) {
 }
+
+// IsAbs is same as filepath.IsAbs on Unix.
+func IsAbs(path string) bool {
```
nit: I think we can simply set ``IsAbs := filepath.IsAbs``, which means that we'd avoid a call on the call stack, similar to how I've done it here: https://github.com/kubernetes/kubernetes/pull/112566/files#diff-2a196acccdccf1b0104a41a86a41a986adff8c975a9d608a9d2dbee96c38c6e9R41

liggitt replied:
I asked for the function to avoid a mutable package variable in an earlier review

claudiubelu commented on pkg/kubelet/apis/config/validation/validation.go:
```diff
@@ -293,11 +295,11 @@ func ValidateKubeletConfiguration(kc *kubeletconfig.KubeletConfiguration, featur
 	}
 
-	if filepath.Clean(kc.PodLogsDir) != kc.PodLogsDir {
+	if runtime.GOOS != "windows" && filepath.Clean(kc.PodLogsDir) != kc.PodLogsDir {
```
hm, I assume we're skipping this because ``filepath.Clean`` replaces ``/`` with ``\``, so this check would fail. But shouldn't we do this check in some form? The check is meant to catch cases in which the configuration is wrong, so there may be some value in doing it.

liggitt replied:
agreed

claudiubelu replied:
Looked a bit into this. What about this:

```
filepath.ToSlash(filepath.Clean(path)) == filepath.ToSlash(path)
```

I've tested this myself on a Windows node:

```
package main

import (
        "fmt"
        "path/filepath"
)

func testCleanPath(path string) bool {
        // Clean will replace slashes to Separator (which is OS-specific).
        // We're making sure that all slashes are the same before comparing.
        return filepath.ToSlash(filepath.Clean(path)) == filepath.ToSlash(path)
}

func main() {
        cases := []struct {
                path     string
                expected bool
        }{
                {
                        path:     "/var/lib/something",
                        expected: true,
                },
                {
                        path:     "C:/var/lib/something",
                        expected: true,
                },
                {
                        path:     "C:\\var\\lib\\something",
                        expected: true,
                },
                {
                        path:     "var/lib/something",
                        expected: true,
                },
                {
                        path:     "var\\lib\\something",
                        expected: true,
                },
                {
                        path:     "/",
                        expected: true,
                },
                {
                        path:     "\\",
                        expected: true,
                },
                {
                        path:     "C:/",
                        expected: true,
                },
                {
                        path:     "C:\\",
                        expected: true,
                },
                {
                        path:     ".",
                        expected: true,
                },
                {
                        path:     "/var/../something",
                        expected: false,
                },
                {
                        path:     "C:\\var\\..\\something",
                        expected: false,
                },
                {
                        path:     "/var//lib/something",
                        expected: false,
                },
                {
                        path:     "C:/var//lib/something",
                        expected: false,
                },
                {
                        path:     "\\var\\\\lib\\something",
                        expected: false,
                },
                {
                        path:     "C:\\var\\\\lib\\something",
                        expected: false,
                },
                {
                        path:     "/var/./lib/something",
                        expected: false,
                },
                {
                        path:     "C:/var/./lib/something",
                        expected: false,
                },
                {
                        path:     "\\var\\.\\lib\\something",
                        expected: false,
                },
                {
                        path:     "C:\\var\\.\\lib\\something",
                        expected: false,
                },
        }

        errorCount := 0
        for _, tc := range cases {
                actual := testCleanPath(tc.path)
                if actual != tc.expected {
                        fmt.Printf("Actual: %q, expected: %q, for path: %s\n", actual, tc.expected, tc.path)
                        errorCount++
                }
        }

        fmt.Printf("Encountered %d errors.\n", errorCount)
}
```

Output from running it:

```
PS C:\Users\Administrator\workdir\kubernetes> go run .\tester.go
Encountered 0 errors.
```

liggitt replied:
that seems reasonable to me

aravindhp replied:
@mxpv what are your thoughts on @claudiubelu's suggestion? Just want to call out that this issue is a blocker for RC2 and the release, so we need to reach a resolution and have this merged soon.

rphillips replied:
I think this makes sense as well. Perhaps adding a IsCleanPath(str) function into utilfs?

mxpv replied:
@aravindhp sure, I will update the PR shortly.

pacoxu commented:
A reminder📢:
- according to slack thread https://kubernetes.slack.com/archives/C0SJ4AFB7/p1712051407401399, this will be a blocker for v1.30.0-rc.2(planned in Thursday, April 11th).

mxpv committed:
Implement IsPathClean to validate path

Signed-off-by: Maksym Pavlenko <pavlenko.maksym@gmail.com>

rphillips commented:
/test pull-kubernetes-e2e-capz-windows-master

rphillips commented:
/lgtm

dims commented:
/approve
/lgtm

aravindhp commented:
/test pull-ci-kubernetes-unit-windows

liggitt commented on pkg/util/filesystem/util_windows_test.go:
```diff
@@ -89,3 +89,12 @@ func TestPendingUnixDomainSocket(t *testing.T) {
+func TestAbsWithSlash(t *testing.T) {
+	// On Windows, filepath.IsAbs will not return True for paths prefixed with a slash
+	assert.True(t, IsAbs("/test"))
+	assert.True(t, IsAbs("\\test"))
```
Is a UNC path really absolute? If we're narrowly trying to make absolute unix paths pass validation, should this allow '/…' but not '//…'?

claudiubelu replied:
UNC paths are absolute. It is also treated as such in golang's ``filepath.IsAbs`` [1].

Additionally, if we look at [2], we see:


... A file name is relative to the current directory if it does not begin with one of the following:
- A UNC name of any format, which always start with two backslash characters ("\\"). For more information, see the next section.
- A disk designator with a backslash, for example "C:\" or "d:\".
- A single backslash, for example, "\directory" or "\file.txt". This is also referred to as an absolute path.

According to this, UNCs are not relative.

[1] https://cs.opensource.google/go/go/+/refs/tags/go1.22.2:src/path/filepath/path_windows.go;l=64
[2] https://learn.microsoft.com/en-us/windows/win32/fileio/naming-a-file#fully-qualified-vs-relative-paths

dims replied:
@liggitt is this info enough to resolve this conversation?

liggitt replied:
yeah, if filepath.IsAbs returns true for UNC paths, that's fine

claudiubelu replied:
I've run this on Windows:

```
package main

import (
        "fmt"
        "path/filepath"
)

func main() {
        cases := []string {
                `C:/var/lib/something`,
                `//./var/lib/something`,
                `//?/var/lib/something`,
                `//foo/C$/var/lib/something`,

                `C:\var\lib\something`,
                `\\.\var\lib\something`,
                `\\?\var\lib\something`,
                `\\foo\C$\var\lib\something`,
        }

        for _, path := range cases {
                fmt.Printf("IsAbs returns %t for: %s\n", filepath.IsAbs(path), path)
        }
}
```

Output:

```
PS C:\Users\Administrator\workdir\sys> go run .\tester.go
IsAbs returns true for: C:/var/lib/something
IsAbs returns true for: //./var/lib/something
IsAbs returns true for: //?/var/lib/something
IsAbs returns true for: //foo/C$/var/lib/something
IsAbs returns true for: C:\var\lib\something
IsAbs returns true for: \\.\var\lib\something
IsAbs returns true for: \\?\var\lib\something
IsAbs returns true for: \\foo\C$\var\lib\something
```

liggitt commented on pkg/util/filesystem/util_windows.go:
```diff
@@ -85,3 +87,10 @@ func IsUnixDomainSocket(filePath string) (bool, error) {
 	return true, nil
 }
+
+// IsAbs returns wether the given path is absolute or not.
```
This needs a big warning on it that it isn't safe to use for API values which will propagate across systems (e.g. REST API values that get validated on Unix, persisted, then consumed by Windows, etc)

dims replied:
+1 Jordan. Let's do this @mxpv

mxpv replied:
Updated in cefe0b2ed04806186ecc671474bf2dbece4857ae

liggitt commented on pkg/util/filesystem/util_windows.go:
```diff
@@ -85,3 +87,10 @@ func IsUnixDomainSocket(filePath string) (bool, error) {
 	return true, nil
 }
+
+// IsAbs returns wether the given path is absolute or not.
```
```suggestion
// IsAbs returns whether the given path is absolute or not.
```

mxpv replied:
Fixed in cefe0b2ed04806186ecc671474bf2dbece4857ae

mxpv committed:
Add warn comment for IsAbs

Signed-off-by: Maksym Pavlenko <pavlenko.maksym@gmail.com>

liggitt commented:
/lgtm
/approve

marosset commented:
/lgtm
